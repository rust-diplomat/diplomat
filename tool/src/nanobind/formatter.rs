//! This module contains functions for formatting types

use crate::cpp::Cpp2Formatter;
use diplomat_core::hir::{DocsUrlGenerator, Method, TypeContext, TypeId};
use std::{borrow::Cow, sync::LazyLock};

/// This type mediates all formatting
///
/// All identifiers from the HIR should go through here before being formatted
/// into the output: This makes it easy to handle reserved words or add rename support
///
/// If you find yourself needing an identifier formatted in a context not yet available here, please add a new method
///
/// This type may be used by other backends attempting to figure out the names
/// of C types and methods.
pub(crate) struct PyFormatter<'tcx> {
    pub cxx: Cpp2Formatter<'tcx>,
}

impl<'tcx> PyFormatter<'tcx> {
    pub fn new(tcx: &'tcx TypeContext, docs_url_gen: &'tcx DocsUrlGenerator) -> Self {
        Self {
            cxx: Cpp2Formatter::new(tcx, docs_url_gen),
        }
    }

    pub fn fmt_binding_fn(&self, id: TypeId, namespaced: bool) -> String {
        let resolved = self.cxx.c.tcx().resolve_type(id);
        let type_name = resolved
            .attrs()
            .rename
            .apply(resolved.name().as_str().into());
        match &resolved.attrs().namespace {
            Some(ns) if namespaced => {
                format!("{ns}::add_{type_name}_binding")
            }
            _ => {
                format!("add_{type_name}_binding")
            }
        }
    }

    pub fn fmt_binding_impl_path(&self, id: TypeId) -> String {
        self.cxx.fmt_type_name(id).replace("::", "/") + "_binding.cpp"
    }

    /// Resolve and format the nested module names for this type
    /// Returns an iterator to the namespaces. Will always have at least one entry
    pub fn fmt_namespaces(&self, id: TypeId) -> impl Iterator<Item = &'tcx str> {
        let resolved = self.cxx.c.tcx().resolve_type(id);
        resolved
            .attrs()
            .namespace
            .as_ref()
            .map(|v| v.split("::"))
            .into_iter()
            .flatten()
    }

    pub fn fmt_method_name<'a>(&'tcx self, method: &'a Method) -> Cow<'a, str> {
        self.fmt_identifier(method.attrs.rename.apply(method.name.as_str().into()))
    }

    pub fn fmt_identifier<'a>(&'tcx self, name: Cow<'a, str>) -> Cow<'a, str> {
        // Source https://docs.python.org/3/reference/lexical_analysis.html#keywords
        #[rustfmt::skip]
        static PY_KEYWORDS: LazyLock<std::collections::HashSet<&str>> = LazyLock::new(|| {
            [
                "False", "await", "else", "import", "pass",
                "None", "break", "except", "in", "raise",
                "True", "class", "finally", "is", "return",
                "and", "continue", "for", "lambda", "try",
                "as", "def", "from", "nonlocal", "while",
                "assert", "del", "global", "not", "with",
                "async", "elif", "if", "or", "yield",
            ]
            .into()
        });

        if PY_KEYWORDS.contains(name.as_ref()) {
            format!("{name}_").into()
        } else {
            name
        }
    }
}
