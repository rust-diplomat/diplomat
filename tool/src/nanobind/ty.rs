use super::root_module::RootModule;
use super::PyFormatter;
use crate::{c::TyGenContext as C2TyGenContext, hir, ErrorStore};
use askama::Template;
use diplomat_core::hir::{OpaqueOwner, StructPathLike, TyPosition, Type, TypeId};
use itertools::Itertools;
use std::borrow::Cow;
use std::collections::BTreeMap;
use std::collections::BTreeSet;

/// A type name with a corresponding variable name, such as a struct field or a function parameter.
#[derive(Clone)]
struct NamedType<'a> {
    var_name: Cow<'a, str>,
    type_name: Cow<'a, str>,
}

/// Everything needed for rendering a method.
struct MethodInfo<'a> {
    /// HIR of the method being rendered
    method: &'a hir::Method,
    /// The python method name
    method_name: Cow<'a, str>,
    /// The C++ method name. May differ due to keyword renaming or other constraints.
    cpp_method_name: Cow<'a, str>,
    // def statement to use - def, def_static, def_prop_ro, etc
    def: String,
    /// The property name, if any
    prop_name: Option<Cow<'a, str>>,
    // If this is a property, this is the associated setter's c++ method name
    setter_name: Option<Cow<'a, str>>,
    // In the *rare* event that they're required, the C++ names & types of the function params
    param_decls: Option<Vec<NamedType<'a>>>,
    // The lifetime annotation required for the method, if any. May be keep_alive<...> or reference_internal
    // Everything else is handled by the automatic behavior depending on return type.
    lifetime_args: Option<Cow<'a, str>>,
}

/// Context for generating a particular type's impl
/// 'tcx refers to the lifetime of the typecontext
/// 'cx refers to the lifetime of the context itself
pub(super) struct TyGenContext<'cx, 'tcx> {
    pub formatter: &'cx PyFormatter<'tcx>,
    pub errors: &'cx ErrorStore<'tcx, String>,
    pub c2: C2TyGenContext<'cx, 'tcx>,
    pub root_module: &'cx mut RootModule<'tcx>,
    pub submodules: &'cx mut BTreeMap<Cow<'tcx, str>, BTreeSet<Cow<'tcx, str>>>,
    pub includes: &'cx mut BTreeSet<String>,
    /// Are we currently generating struct fields?
    pub generating_struct_fields: bool,
}

impl<'ccx, 'tcx: 'ccx> TyGenContext<'ccx, 'tcx> {
    /// Checks for & adds modules with their parents to the root module definition.
    pub fn gen_modules(&mut self, id: TypeId, _docstring: Option<&str>) {
        let namespaces = self.formatter.fmt_namespaces(id);

        let mut parent = self.root_module.module_name.clone();
        for module in namespaces {
            println!("Adding {module} to parent {parent}");
            self.submodules
                .entry(parent)
                .or_default()
                .insert(module.into());
            parent = module.into();
        }
    }

    /// Adds an enum definition to the current implementation.
    ///
    /// The enum is defined in C++ using a `class` with a single private field that is the
    /// C enum type. This enables us to add methods to the enum and generally make the enum
    /// behave more like an upgraded C++ type. We don't use `enum class` because methods
    /// cannot be added to it.
    pub fn gen_enum_def<W: std::fmt::Write + ?Sized>(
        &mut self,
        ty: &'tcx hir::EnumDef,
        id: TypeId,
        out: &mut W,
    ) {
        let type_name = self.formatter.cxx.fmt_type_name(id);
        let type_name_unnamespaced = self.formatter.cxx.fmt_type_name_unnamespaced(id);

        let values = ty
            .variants
            .iter()
            .map(|e| self.formatter.cxx.fmt_enum_variant(e))
            .collect::<Vec<_>>();

        #[derive(Template)]
        #[template(path = "nanobind/enum_impl.cpp.jinja", escape = "none")]
        struct ImplTemplate<'a> {
            type_name: &'a str,
            values: Vec<Cow<'a, str>>,
            type_name_unnamespaced: &'a str,
        }

        ImplTemplate {
            type_name: &type_name,
            values,
            type_name_unnamespaced: &type_name_unnamespaced,
        }
        .render_into(out)
        .unwrap();

        self.add_to_root_module(id);
    }

    pub fn add_to_root_module(&mut self, id: TypeId) {
        self.gen_modules(id, None);
        self.root_module
            .fwd_decls
            .entry(self.formatter.fmt_namespaces(id).join("::"))
            .or_default()
            .push(format!(
                "void {}(nb::handle);",
                self.formatter.fmt_binding_fn(id, false)
            ));

        let module_namespaces = [self.root_module.module_name.to_string()]
            .into_iter()
            .chain(self.formatter.fmt_namespaces(id).map(|s| s.to_owned()))
            .collect();

        let entry = self
            .root_module
            .module_fns
            .entry(module_namespaces)
            .or_default();

        entry.push(self.formatter.fmt_binding_fn(id, true));
    }

    pub fn gen_opaque_def<W: std::fmt::Write + ?Sized>(
        &mut self,
        ty: &'tcx hir::OpaqueDef,
        id: TypeId,
        out: &mut W,
    ) {
        let type_name = self.formatter.cxx.fmt_type_name(id);
        let type_name_unnamespaced = self.formatter.cxx.fmt_type_name_unnamespaced(id);

        let methods = self.gen_all_method_infos(id, ty.methods.iter());

        #[derive(Template)]
        #[template(path = "nanobind/opaque_impl.cpp.jinja", escape = "none")]
        struct ImplTemplate<'a> {
            type_name: &'a str,
            methods: &'a [MethodInfo<'a>],
            type_name_unnamespaced: &'a str,
        }

        ImplTemplate {
            type_name: &type_name,
            methods: methods.as_slice(),
            type_name_unnamespaced: &type_name_unnamespaced,
        }
        .render_into(out)
        .unwrap();
        self.add_to_root_module(id);
    }

    pub fn gen_struct_def<P: TyPosition, W: std::fmt::Write + ?Sized>(
        &mut self,
        def: &'tcx hir::StructDef<P>,
        id: TypeId,
        out: &mut W,
    ) {
        let type_name = self.formatter.cxx.fmt_type_name(id);
        let type_name_unnamespaced = self.formatter.cxx.fmt_type_name_unnamespaced(id);

        self.generating_struct_fields = true;
        let field_decls = def
            .fields
            .iter()
            .map(|field| self.gen_ty_decl(&field.ty, field.name.as_str()))
            .collect::<Vec<_>>();
        self.generating_struct_fields = false;

        let methods = self.gen_all_method_infos(id, def.methods.iter());

        self.gen_modules(id, None);
        #[derive(Template)]
        #[template(path = "nanobind/struct_impl.cpp.jinja", escape = "none")]
        struct ImplTemplate<'a> {
            type_name: &'a str,
            fields: &'a [NamedType<'a>],
            methods: &'a [MethodInfo<'a>],
            type_name_unnamespaced: &'a str,
            has_constructor: bool,
        }

        ImplTemplate {
            type_name: &type_name,
            fields: field_decls.as_slice(),
            methods: methods.as_slice(),
            type_name_unnamespaced: &type_name_unnamespaced,
            has_constructor: methods.iter().any(|v| {
                matches!(
                    v.method.attrs.special_method,
                    Some(hir::SpecialMethod::Constructor)
                )
            }),
        }
        .render_into(out)
        .unwrap();
        self.add_to_root_module(id);
    }

    fn gen_all_method_infos(
        &mut self,
        id: TypeId,
        methods: std::slice::Iter<'tcx, hir::Method>,
    ) -> Vec<MethodInfo<'ccx>> {
        // BTree map ensures that the output will be sorted by method name for more consistent codegen output.
        let mut method_infos = BTreeMap::<String, MethodInfo>::new();

        for method in methods {
            if let Some(info) = self.gen_method_info(id, method) {
                method_infos
                    // Use the property name as the key if present so we can collapse getters & setters
                    .entry(info.prop_name.clone().unwrap_or(info.method_name.clone()).to_string())
                    .and_modify(|e| {
                        // If the entry already exists...
                        match method.attrs.special_method {
                            Some(hir::SpecialMethod::Getter(_)) => {
                                assert!(
                                    matches!(e.method.attrs.special_method, Some(hir::SpecialMethod::Setter(_))),
                                    "Method Info entry for {} already exists, but isn't a setter.",
                                    e.method_name
                                );
                                // If the entry previously associated with this was the setter,
                                // overwrite the method reference & method name with this getter
                                e.method = method;
                                e.method_name = info.method_name.clone();
                                e.cpp_method_name = info.cpp_method_name.clone();
                            }
                            Some(hir::SpecialMethod::Setter(_)) => {
                                assert!(
                                    matches!(
                                        e.method.attrs.special_method,
                                        Some(hir::SpecialMethod::Getter(_))
                                    ),
                                    "Method Info entry for {} already exists, but isn't a getter.",
                                    e.method_name
                                );
                                e.setter_name = Some(info.method_name.clone());
                                e.def = info.def.clone(); // when a setter exists, use it's qualifiers instead.
                                e.param_decls = info.param_decls.clone(); // also it's params, since the getter has none by definition.
                            }
                            _ => { panic!("Method Info for {} already exists but isn't a getter or setter!", e.method_name); }
                        };
                    })
                    .or_insert(info);
            }
        }

        for info in method_infos.values() {
            if matches!(
                info.method.attrs.special_method,
                Some(hir::SpecialMethod::Setter(_))
            ) {
                self.errors.push_error(format!(
                    "Setter {} exists without a matching getter, which is not allowed by nanobind.",
                    info.method_name
                ));
            }
        }

        method_infos.into_values().collect()
    }

    fn gen_method_info(
        &mut self,
        id: TypeId,
        method: &'tcx hir::Method,
    ) -> Option<MethodInfo<'ccx>> {
        if method.attrs.disable {
            return None;
        }
        let _guard = self.errors.set_context_method(
            self.c2.tcx.fmt_type_name_diagnostics(id),
            method.name.as_str().into(),
        );
        let cpp_method_name = self.formatter.cxx.fmt_method_name(method);
        let method_name = self.formatter.fmt_method_name(method);
        let mut setter_name = None;

        let mut def_qualifiers = vec!["def"];

        let mut prop_name = None;
        if let Some(hir::SpecialMethod::Getter(name)) = &method.attrs.special_method {
            def_qualifiers.extend(["prop_ro"]);
            prop_name = Some(
                name.as_ref()
                    .map(|v| v.into())
                    .unwrap_or(method_name.clone()),
            );
        } else if let Some(hir::SpecialMethod::Setter(name)) = &method.attrs.special_method {
            def_qualifiers.extend(["prop_rw"]);
            setter_name = Some(method_name.clone());
            prop_name = Some(
                name.as_ref()
                    .map(|v| v.into())
                    .unwrap_or(method_name.clone()),
            );
        }

        if method.param_self.is_none()
            && !matches!(
                method.attrs.special_method,
                Some(hir::SpecialMethod::Constructor) // Constructors weirdly don't use def_static
            )
        {
            def_qualifiers.extend(["static"]);
        }

        let param_decls = {
            if matches!(
                method.attrs.special_method,
                Some(hir::SpecialMethod::Constructor) | Some(hir::SpecialMethod::Setter(_)) // We only need type info for constructors or certain setters
            ) && !matches!(
                // and even then, only when the type isn't opaque
                id,
                TypeId::Opaque(_)
            ) {
                Some(
                    method
                        .params
                        .iter()
                        .map(|p| NamedType {
                            var_name: self.formatter.cxx.fmt_param_name(p.name.as_str()),
                            type_name: self.gen_type_name(&p.ty),
                        })
                        .collect(),
                )
            } else {
                None
            }
        };

        let mut visitor = method.borrowing_param_visitor(self.c2.tcx, false);

        // Collect all the relevant borrowed params, with self in position 1 if present
        let mut param_borrows = method
            .param_self
            .iter()
            .map(|s| visitor.visit_param(&s.ty.clone().into(), "self"))
            .collect::<Vec<_>>();
        // Must be a separate call *after* collect to avoid double-borrowing visitor
        param_borrows.extend(
            method
                .params
                .iter()
                .map(|p| visitor.visit_param(&p.ty, p.name.as_str())),
        );

        let self_number = if matches!(
            method.attrs.special_method,
            Some(hir::SpecialMethod::Constructor)
        ) {
            1 // per the docs, constructors don't have "returns", they act on "self"
        } else {
            0
        };

        let mut lifetime_args = vec![];

        // No keep_alive for even borrowed string outputs, the type conversion always involves a copy
        if !matches!(
            method.output.success_type(),
            hir::SuccessType::OutType(hir::Type::Slice(hir::Slice::Str(..)))
        ) {
            lifetime_args.extend(
                param_borrows
                    .into_iter()
                    .enumerate()
                    .filter_map(|(i, p)| match p {
                        hir::borrowing_param::ParamBorrowInfo::BorrowedSlice
                        | hir::borrowing_param::ParamBorrowInfo::Struct(_)
                        | hir::borrowing_param::ParamBorrowInfo::BorrowedOpaque => {
                            Some(format!(
                                "nb::keep_alive<{self_number}, {}>()",
                                i + 1 + self_number
                            )) // Keep 0 (the return) alive until the element at P is returned
                        }
                        _ => None,
                    })
                    .collect::<Vec<_>>(),
            );
        }

        if matches!(method.output.success_type(), hir::SuccessType::OutType(hir::Type::Opaque(path)) if !path.is_owned())
        {
            lifetime_args.push("nb::rv_policy::reference".to_owned());
        }

        Some(MethodInfo {
            method,
            method_name,
            cpp_method_name,
            def: def_qualifiers.join("_"),
            setter_name,
            prop_name,
            param_decls,
            lifetime_args: if lifetime_args.is_empty() {
                None
            } else {
                Some(lifetime_args.join(", ").into())
            },
        })
    }

    /// Generates C++ code for referencing a particular type with a given name.
    fn gen_ty_decl<'a, P: TyPosition>(&mut self, ty: &Type<P>, var_name: &'a str) -> NamedType<'a>
    where
        'ccx: 'a,
    {
        let var_name = self.formatter.cxx.fmt_param_name(var_name);
        let type_name = self.gen_type_name(ty);

        NamedType {
            var_name,
            type_name,
        }
    }

    /// Generates Python code for referencing a particular type.
    ///
    /// This function adds the necessary type imports to the decl and impl files.
    fn gen_type_name<P: TyPosition>(&mut self, ty: &Type<P>) -> Cow<'ccx, str> {
        match *ty {
            Type::Primitive(prim) => self.formatter.cxx.fmt_primitive_as_c(prim),
            Type::Opaque(ref op) => {
                let op_id = op.tcx_id.into();
                let type_name = self.formatter.cxx.fmt_type_name(op_id);
                let def = self.c2.tcx.resolve_type(op_id);

                if def.attrs().disable {
                    self.errors
                        .push_error(format!("Found usage of disabled type {type_name}"))
                }
                let mutability = op.owner.mutability().unwrap_or(hir::Mutability::Mutable);
                let ret = match (op.owner.is_owned(), op.is_optional()) {
                    // unique_ptr is nullable
                    (true, _) => self.formatter.cxx.fmt_owned(&type_name),
                    (false, true) => self
                        .formatter
                        .cxx
                        .fmt_optional_borrowed(&type_name, mutability),
                    (false, false) => self.formatter.cxx.fmt_borrowed(&type_name, mutability),
                };
                let ret = ret.into_owned().into();

                self.includes
                    .insert(self.formatter.cxx.fmt_impl_header_path(op_id));
                ret
            }
            Type::Struct(ref st) => {
                let id = st.id();
                let type_name = self.formatter.cxx.fmt_type_name(id);
                let def = self.c2.tcx.resolve_type(id);
                if def.attrs().disable {
                    self.errors
                        .push_error(format!("Found usage of disabled type {type_name}"))
                }

                self.includes
                    .insert(self.formatter.cxx.fmt_impl_header_path(id));
                type_name
            }
            Type::Enum(ref e) => {
                let id = e.tcx_id.into();
                let type_name = self.formatter.cxx.fmt_type_name(id);
                let def = self.c2.tcx.resolve_type(id);
                if def.attrs().disable {
                    self.errors
                        .push_error(format!("Found usage of disabled type {type_name}"))
                }

                self.includes
                    .insert(self.formatter.cxx.fmt_impl_header_path(id));
                type_name
            }
            Type::Slice(hir::Slice::Str(_, encoding)) => {
                self.formatter.cxx.fmt_borrowed_str(encoding)
            }
            Type::Slice(hir::Slice::Primitive(b, p)) => {
                let ret = self.formatter.cxx.fmt_primitive_as_c(p);
                let ret = self.formatter.cxx.fmt_borrowed_slice(
                    &ret,
                    b.map(|b| b.mutability).unwrap_or(hir::Mutability::Mutable),
                );
                ret.into_owned().into()
            }
            Type::Slice(hir::Slice::Strs(encoding)) => format!(
                "diplomat::span<const {}>",
                self.formatter.cxx.fmt_borrowed_str(encoding)
            )
            .into(),
            Type::DiplomatOption(ref inner) => {
                format!("std::optional<{}>", self.gen_type_name(inner)).into()
            }
            Type::Callback(..) => "".into(),
            _ => unreachable!("unknown AST/HIR variant"),
        }
    }
}
