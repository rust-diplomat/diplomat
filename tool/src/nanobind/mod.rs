mod binding;
mod formatter;
mod ty;

use std::{borrow::Cow, collections::HashSet};

use crate::{Config, ErrorStore, FileMap};
use binding::Binding;
use diplomat_core::hir::{self, BackendAttrSupport, DocsUrlGenerator};
use formatter::PyFormatter;
use serde::{Deserialize, Serialize};
use ty::TyGenContext;

use crate::cpp;

// Python support using the nanobind c++ library to create a python binding.
//
// The generated nanobind.cpp files requires linking with nanobind
// See the feature_test project for an example of a pyproject & CMakeLists.txt which can be compiled
// using pip install. Compilation requires a C++ compiler & CMake, as well as a downloaded
// copy of libnanobind.

pub(crate) fn attr_support() -> BackendAttrSupport {
    let mut a = BackendAttrSupport::default();

    a.namespacing = true;
    a.memory_sharing = true;
    a.non_exhaustive_structs = false;
    a.method_overloading = true;
    a.utf8_strings = true;
    a.utf16_strings = true;
    a.static_slices = true;

    a.constructors = true;
    a.named_constructors = false;
    a.fallible_constructors = true;
    a.accessors = true;
    a.static_accessors = true;
    a.comparators = true;
    a.stringifiers = true;
    a.iterators = true;
    a.iterables = true;
    a.arithmetic = true;
    a.indexing = true;
    a.option = true;
    a.callbacks = true;
    a.traits = false;

    a
}

#[derive(Debug, Clone, Deserialize, Serialize)]
struct PythonConfig {
    lib_name: String,
}

pub(crate) fn run<'cx>(
    tcx: &'cx hir::TypeContext,
    conf: Config,
    docs: &'cx DocsUrlGenerator,
) -> (FileMap, ErrorStore<'cx, String>) {
    let files = FileMap::default();
    let formatter = PyFormatter::new(tcx, docs);
    let errors = ErrorStore::default();

    let lib_name = conf
        .shared_config
        .lib_name
        .expect("Nanobind backend requires lib_name to be set in the config");

    // Output the C++ bindings we rely on

    let (cpp_files, cpp_errors) = cpp::run(tcx, docs);

    files.files.borrow_mut().extend(
        cpp_files
            .files
            .take()
            .into_iter()
            .map(|(k, v)| (format!("include/{k}"), v)),
    );
    errors.errors.borrow_mut().extend(cpp_errors.errors.take());

    let nanobind_filepath = format!("{lib_name}_ext.cpp");
    let mut binding = Binding::new();
    binding.module_name = lib_name.into();

    let mut submodules = HashSet::<Cow<str>>::new();
    for (id, ty) in tcx.all_types() {
        if ty.attrs().disable {
            // Skip type if disabled
            continue;
        }

        let decl_header_path = formatter.cxx.fmt_decl_header_path(id);
        let impl_file_path = formatter.cxx.fmt_impl_header_path(id);

        let mut context = TyGenContext {
            formatter: &formatter,
            errors: &errors,
            c2: crate::c::TyGenContext {
                tcx,
                formatter: &formatter.cxx.c,
                errors: &errors,
                is_for_cpp: false,
                id: id.into(),
                decl_header_path: &decl_header_path,
                impl_header_path: &impl_file_path,
            },
            binding: &mut binding,
            submodules: &mut submodules,
            generating_struct_fields: false,
        };

        context
            .binding
            .includes
            .insert(impl_file_path.clone().into());

        let guard = errors.set_context_ty(ty.name().as_str().into());
        match ty {
            hir::TypeDef::Enum(o) => context.gen_enum_def(o, id),
            hir::TypeDef::Opaque(o) => context.gen_opaque_def(o, id),
            hir::TypeDef::Struct(s) => context.gen_struct_def(s, id),
            hir::TypeDef::OutStruct(s) => context.gen_struct_def(s, id),
            _ => unreachable!("unknown AST/HIR variant"),
        }
        drop(guard);
    }
    files.add_file(nanobind_filepath.to_owned(), binding.to_string());

    (files, errors)
}

#[cfg(test)]
mod test {
    use diplomat_core::hir::{self, TypeDef};
    use quote::quote;
    use std::borrow::Cow;
    use std::collections::HashSet;

    #[test]
    fn test_opaque_gen() {
        let tokens = quote! {
            #[diplomat::bridge]
            #[diplomat::attr(auto, namespace = "mylib")]
            mod ffi {

                #[diplomat::opaque]
                struct OpaqueStruct;

                impl OpaqueStruct {
                    pub fn new() -> Box<OpaqueStruct> {
                        Box::new(OpaqueStruct{})
                    }

                    pub fn do_thing() -> bool {
                        return true;
                    }
                }
            }
        };
        let item = syn::parse2::<syn::File>(tokens).expect("failed to parse item ");

        let mut attr_validator = hir::BasicAttributeValidator::new("python");
        attr_validator.support = crate::nanobind::attr_support();

        let tcx = match hir::TypeContext::from_syn(&item, attr_validator) {
            Ok(context) => context,
            Err(e) => {
                for (_cx, err) in e {
                    eprintln!("Lowering error: {}", err);
                }
                panic!("Failed to create context")
            }
        };

        let (type_id, opaque_def) = match tcx
            .all_types()
            .next()
            .expect("Failed to generate first opaque def")
        {
            (type_id, TypeDef::Opaque(opaque_def)) => (type_id, opaque_def),
            _ => panic!("Failed to find opaque type from AST"),
        };

        let docs_gen = Default::default();
        let formatter = crate::nanobind::PyFormatter::new(&tcx, &docs_gen);
        let errors = crate::ErrorStore::default();
        let mut binding = crate::nanobind::Binding::new();
        binding.module_name = std::borrow::Cow::Borrowed("pymod");

        let decl_header_path = formatter.cxx.fmt_decl_header_path(type_id);
        let impl_file_path = formatter.cxx.fmt_impl_header_path(type_id);

        let mut submodules = HashSet::<Cow<str>>::new();
        let mut context = crate::nanobind::TyGenContext {
            formatter: &formatter,
            errors: &errors,
            c2: crate::c::TyGenContext {
                tcx: &tcx,
                formatter: &formatter.cxx.c,
                errors: &errors,
                is_for_cpp: false,
                id: type_id.into(),
                decl_header_path: &decl_header_path,
                impl_header_path: &impl_file_path,
            },
            binding: &mut binding,
            generating_struct_fields: false,
            submodules: &mut submodules,
        };

        context.gen_opaque_def(opaque_def, type_id);
        let generated = binding.to_string();
        insta::assert_snapshot!(generated)
    }

    #[test]
    fn test_enum_gen() {
        let tokens = quote! {
            #[diplomat::bridge]
            #[diplomat::attr(auto, namespace = "mylib")]
            mod ffi {

                #[diplomat::enum_convert(my_thingy::SpeedSetting)]
                pub enum SpeedSetting {
                    Fast, Medium, Slow
                }
            }
        };
        let item = syn::parse2::<syn::File>(tokens).expect("failed to parse item ");

        let mut attr_validator = hir::BasicAttributeValidator::new("python");
        attr_validator.support = crate::nanobind::attr_support();

        let tcx = match hir::TypeContext::from_syn(&item, attr_validator) {
            Ok(context) => context,
            Err(e) => {
                for (_cx, err) in e {
                    eprintln!("Lowering error: {}", err);
                }
                panic!("Failed to create context")
            }
        };

        let (type_id, enum_def) = match tcx
            .all_types()
            .next()
            .expect("Failed to generate first opaque def")
        {
            (type_id, TypeDef::Enum(enum_def)) => (type_id, enum_def),
            _ => panic!("Failed to find opaque type from AST"),
        };

        let docs_gen = Default::default();
        let formatter = crate::nanobind::PyFormatter::new(&tcx, &docs_gen);
        let errors = crate::ErrorStore::default();
        let mut binding = crate::nanobind::Binding::new();
        binding.module_name = std::borrow::Cow::Borrowed("pymod");

        let decl_header_path = formatter.cxx.fmt_decl_header_path(type_id);
        let impl_file_path = formatter.cxx.fmt_impl_header_path(type_id);

        let mut submodules = HashSet::<Cow<str>>::new();
        let mut context = crate::nanobind::TyGenContext {
            formatter: &formatter,
            errors: &errors,
            c2: crate::c::TyGenContext {
                tcx: &tcx,
                formatter: &formatter.cxx.c,
                errors: &errors,
                is_for_cpp: false,
                id: type_id.into(),
                decl_header_path: &decl_header_path,
                impl_header_path: &impl_file_path,
            },
            binding: &mut binding,
            generating_struct_fields: false,
            submodules: &mut submodules,
        };

        context.gen_enum_def(enum_def, type_id);
        let generated = binding.to_string();
        insta::assert_snapshot!(generated)
    }

    #[test]
    fn test_struct_gen() {
        let tokens = quote! {
            #[diplomat::bridge]
            #[diplomat::attr(auto, namespace = "mylib")]
            mod ffi {
                pub struct Thingy {
                    pub a: bool,
                    pub b: u8,
                    pub c: f64,
                }
            }
        };
        let item = syn::parse2::<syn::File>(tokens).expect("failed to parse item ");

        let mut attr_validator = hir::BasicAttributeValidator::new("python");
        attr_validator.support = crate::nanobind::attr_support();

        let tcx = match hir::TypeContext::from_syn(&item, attr_validator) {
            Ok(context) => context,
            Err(e) => {
                for (_cx, err) in e {
                    eprintln!("Lowering error: {}", err);
                }
                panic!("Failed to create context")
            }
        };

        let (type_id, struct_def) = match tcx
            .all_types()
            .next()
            .expect("Failed to generate first opaque def")
        {
            (type_id, TypeDef::Struct(struct_def)) => (type_id, struct_def),
            _ => panic!("Failed to find opaque type from AST"),
        };

        let docs_gen = Default::default();
        let formatter = crate::nanobind::PyFormatter::new(&tcx, &docs_gen);
        let errors = crate::ErrorStore::default();
        let mut binding = crate::nanobind::Binding::new();
        binding.module_name = std::borrow::Cow::Borrowed("pymod");

        let decl_header_path = formatter.cxx.fmt_decl_header_path(type_id);
        let impl_file_path = formatter.cxx.fmt_impl_header_path(type_id);

        let mut submodules = HashSet::<Cow<str>>::new();
        let mut context = crate::nanobind::TyGenContext {
            formatter: &formatter,
            errors: &errors,
            c2: crate::c::TyGenContext {
                tcx: &tcx,
                formatter: &formatter.cxx.c,
                errors: &errors,
                is_for_cpp: false,
                id: type_id.into(),
                decl_header_path: &decl_header_path,
                impl_header_path: &impl_file_path,
            },
            binding: &mut binding,
            generating_struct_fields: false,
            submodules: &mut submodules,
        };

        context.gen_struct_def(struct_def, type_id);
        let generated = binding.to_string();
        insta::assert_snapshot!(generated)
    }
}
