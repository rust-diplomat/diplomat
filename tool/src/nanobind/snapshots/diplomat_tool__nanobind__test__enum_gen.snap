---
source: tool/src/nanobind/mod.rs
expression: enum_gen
---
nb::class_<mylib::SpeedSetting> e_class(mod, "SpeedSetting");

	nb::enum_<mylib::SpeedSetting::Value>(e_class, "SpeedSetting")
		.value("Fast", mylib::SpeedSetting::Fast)
		.value("Medium", mylib::SpeedSetting::Medium)
		.value("Slow", mylib::SpeedSetting::Slow)
		.export_values();

	e_class
		.def(nb::init_implicit<mylib::SpeedSetting::Value>())
		.def(nb::self == mylib::SpeedSetting::Value())
		.def("__repr__", [](const mylib::SpeedSetting& self){
			return nb::str(nb::cast(mylib::SpeedSetting::Value(self)));
		});
